src/rvsdg/tree_unique/to_tree.rs

```diff
@@ -7,3 +7,245 @@
 //! These shared nodes need to be let-bound so that they are only
 //! computed once in the tree encoded
 //! program.
+#[cfg(test)]
+use crate::{cfg::program_to_cfg, rvsdg::cfg_to_rvsdg, util::parse_from_string};
+#[cfg(test)]
+use tree_unique_args::ast::{program, sequence};
+
+use crate::rvsdg::{BasicExpr, Id, Operand, RvsdgBody, RvsdgFunction, RvsdgProgram};
+use bril_rs::{Literal, ValueOps};
+use hashbrown::HashMap;
+use tree_unique_args::{
+    ast::{
+        add, arg, concat, function, get, num, print, program_vec, sequence_vec, tfalse, tlet, ttrue,
+    },
+    Expr,
+};
+
+impl RvsdgProgram {
+    pub fn to_tree_encoding(&self) -> Expr {
+        program_vec(
+            self.functions
+                .iter()
+                .map(|f| f.to_tree_encoding())
+                .collect(),
+        )
+    }
+}
+
+struct RegionTranslator<'a> {
+    /// The number of arguments to this region
+    num_args: usize,
+    /// a stack of let bindings to generate
+    bindings: Vec<Expr>,
+    /// A map from the rvsdg node id
+    /// to the index in the argument
+    /// where the value is stored.
+    /// After evaluating a node, do not evaluate it again.
+    /// Instead find it's index here.
+    index_of: HashMap<Id, usize>,
+    nodes: &'a Vec<RvsdgBody>,
+}
+
+fn cbind(expr: Expr, body: Expr) -> Expr {
+    tlet(concat(arg(), expr), body)
+}
+
+impl<'a> RegionTranslator<'a> {
+    /// Adds a binding and returns its index
+    /// into the argument list.
+    fn add_binding(&mut self, expr: Expr, id: Id) -> usize {
+        self.bindings.push(expr);
+        let res = self.bindings.len() - 1 + self.num_args;
+        assert_eq!(self.index_of.insert(id, res), None);
```

Comment:
What is this assertion?

Reply:
(Could use a comment explaining what the invariant is / why it must hold / why it matters

Reply:
Added an error message that describes why it shouldn't occur